contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,957 @@
+     * Scenario 3: When there is strategy's liquidity in the pool all 3 steps are taken
+     *
+     *
+     * Exact _amountToSwap, _minTokenReceived & _swapWeth parameters shall be determined by simulating the
+     * transaction off-chain. The strategy checks that after the swap the share of the tokens is in the
+     * expected ranges.
+     *
+     * @param _amountToSwap The amount of the token to swap
+     * @param _minTokenReceived Slippage check -> minimum amount of token expected in return
+     * @param _swapWeth Swap using WETH when true, use OETHb when false
+     */
+    function rebalance(
+        uint256 _amountToSwap,
+        uint256 _minTokenReceived,
+        bool _swapWeth
+    ) external nonReentrant onlyGovernorOrStrategist {
+        _rebalance(_amountToSwap, _minTokenReceived, _swapWeth);
+    }
+
+    /**
+     * @dev Remove almost all of the liqudity, rebalance the pool to the desired token split and
+     * deposit all of the liquidity.
+     */
+    // rebalance already has reentrancy check
+    // slither-disable-start reentrancy-no-eth
+    function _rebalance(
+        uint256 _amountToSwap,
+        uint256 _minTokenReceived,
+        bool _swapWeth
+    ) internal {
+        /**
+         * Would be nice to check if there is any total liquidity in the pool before performing this swap
+         * but there is no easy way to do that in UniswapV3:
+         * - clPool.liquidity() -> only liquidity in the active tick
+         * - asset[1&2].balanceOf(address(clPool)) -> will include uncollected tokens of LP providers
+         *   after their liquidity position has been decreased
+         */
+
+        /**
+         * When rebalance is called for the first time or after a withdrawAll there is no strategy's
+         * liquidity in the pool yet. The partial removal is thus skipped.
+         */
+        if (tokenId != 0) {
+            _removePartialLiquidity(withdrawLiquidityShare);
+        }
+        // in some cases we will just want to add liquidity and not issue a swap to move the
+        // active trading position within the pool
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _minTokenReceived, _swapWeth);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        if (tokenId != 0) {
+            _checkLiquidityWithinExpectedShare();
+        }
+
+        _addLiquidity();
+        _checkLiquidityWithinExpectedShare();
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /**
+     * @dev Decrease partial liquidity from the pool.
+     * @param _partialLiquidityToDecrease The amount of liquidity to remove expressed in 18 decimals
+     */
+    function _removePartialLiquidity(uint256 _partialLiquidityToDecrease)
+        internal
+    {
+        require(_partialLiquidityToDecrease > 0, "Must remove some liquidity");
+        require(
```

Comment:
Why is this a requirement? Removing all liquidity should be okay, right? And even desirable?

Reply:
🟠 For that matter, why do we need this method? It's just wrapping `_removeLiquidity` that should do these checks anyway.

Reply:
These functions are the result of a decision that the rebalance function will always remove part of the liquidity and not all of it. That is why the remove liquidity functions also have this limitation. If we decide not to impose the limitation of needing to leave some liquidity in during rebalance we could: 
- have only 1 internal function for removal of liquidity and:   
  - have that function handle whether to burn the NFT token or not after removal of liquidity (if there is none ther)
  - or better yet simplify the code and not bother with burning of the NFT token even when there is 0 liquidity.

Reply:
I like having just one function much better. Also like not having the artificial distinction between removing most liquidity and removing all.

Reply:
As mentioned here https://github.com/OriginProtocol/origin-dollar/pull/2184#discussion_r1725035464, we probably want to temporary pull all liquidity before swaps (but without burning the NFT)

Reply:
Changes done in separate PR: https://github.com/OriginProtocol/origin-dollar/pull/2198. This does somewhat further simplify the code which is great.